plugin-server/src/cdp/consumers/cdp-behavioural-events.consumer.ts

```diff
@@ -0,0 +1,203 @@
+import { Message } from 'node-rdkafka'
+import { Counter } from 'prom-client'
+
+import { convertToHogFunctionInvocationGlobals } from '../../cdp/utils'
+import { KAFKA_EVENTS_JSON } from '../../config/kafka-topics'
+import { KafkaConsumer } from '../../kafka/consumer'
+import { runInstrumentedFunction } from '../../main/utils'
+import { Hub, RawClickHouseEvent } from '../../types'
+import { parseJSON } from '../../utils/json-parse'
+import { logger } from '../../utils/logger'
+import { HogFunctionInvocationGlobals } from '../types'
+import { execHog } from '../utils/hog-exec'
+import { convertToHogFunctionFilterGlobal } from '../utils/hog-function-filtering'
+import { CdpConsumerBase } from './cdp-base.consumer'
+
+export const counterParseError = new Counter({
+    name: 'cdp_behavioural_function_parse_error',
+    help: 'A behavioural function invocation was parsed with an error',
+    labelNames: ['error'],
+})
+
+export const counterEventsConsumed = new Counter({
+    name: 'cdp_behavioural_events_consumed_total',
+    help: 'Total number of events consumed by the behavioural consumer',
+})
+
+export const counterEventsMatchedTotal = new Counter({
+    name: 'cdp_behavioural_events_matched_total',
+    help: 'Total number of events that matched at least one action filter',
+})
+
+export class CdpBehaviouralEventsConsumer extends CdpConsumerBase {
+    protected name = 'CdpBehaviouralEventsConsumer'
+    protected kafkaConsumer: KafkaConsumer
+
+    constructor(hub: Hub, topic: string = KAFKA_EVENTS_JSON, groupId: string = 'cdp-behavioural-events-consumer') {
+        super(hub)
+        this.kafkaConsumer = new KafkaConsumer({ groupId, topic })
+    }
+
+    public async processBatch(invocationGlobals: HogFunctionInvocationGlobals[]): Promise<void> {
+        return await this.runInstrumented('processBatch', async () => {
+            if (!invocationGlobals.length) {
+                return
+            }
+
+            // Track events consumed and matched (absolute numbers)
+            let eventsMatched = 0
+
+            for (const event of invocationGlobals) {
+                const matched = await this.processEvent(event)
+                if (matched) {
+                    eventsMatched++
+                }
+            }
+
+            // Update metrics with absolute numbers
+            counterEventsConsumed.inc(invocationGlobals.length)
+            counterEventsMatchedTotal.inc(eventsMatched)
+        })
+    }
+
+    private async processEvent(event: HogFunctionInvocationGlobals): Promise<boolean> {
+        try {
+            const teamId = event.project.id
+            const actions = await this.loadActionsForTeam(teamId)
+
+            if (!actions.length) {
+                logger.debug('No actions found for team', { teamId })
+                return false
+            }
+
+            let eventMatched = false
+            for (const action of actions) {
```

Comment:
Again this can safely be a promise.all - its less of a big deal given that it is all cpu bound work but the execHog thing is safe to do in a proper loop now so there is no actual downside